driver/src/price_estimation/kraken/api.rs

```diff
@@ -0,0 +1,272 @@
+            .context("failed to parse ticker JSON")?
+            .into_result()
+    }
+}
+
+/// The result type that is returned by Kraken on API requests. This type is
+/// only used internally.
+#[derive(Clone, Debug, Deserialize)]
+struct KrakenResult<T> {
+    error: Vec<String>,
+    result: Option<T>,
+}
+
+impl<T> KrakenResult<T> {
+    fn into_result(self) -> Result<T> {
+        if let Some(result) = self.result {
+            Ok(result)
+        } else if !self.error.is_empty() {
+            Err(anyhow!("Kraken API errors: {:?}", self.error))
+        } else {
+            Err(anyhow!("unknown Kraken API error"))
+        }
+    }
+}
+
+/// A struct representing an asset retrieved from the Kraken API.
+///
+/// Note that this is only a small subset of the data provided by the Kraken API
+/// and only the parts required for retrieving price estimates for the solver
+/// are included.
+#[derive(Clone, Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct Asset {
+    pub altname: String,
+}
+
+impl Asset {
+    /// Create a new asset from an alternate name.
+    #[cfg(test)]
+    pub fn new(altname: &str) -> Asset {
+        Asset {
+            altname: altname.into(),
+        }
+    }
+}
+
+/// A struct representing an asset pair retrieved from the Kraken API.
+///
+/// Note that this is only a small subset of the data provided by the Kraken API
+/// and only the parts required for retrieving price estimates for the solver
+/// are included.
+#[derive(Clone, Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct AssetPair {
+    pub base: String,
+    pub quote: String,
+}
+
+impl AssetPair {
+    /// Create a new asset pair from base and quote assets.
+    #[cfg(test)]
+    pub fn new(base: &str, quote: &str) -> AssetPair {
+        AssetPair {
+            base: base.into(),
+            quote: quote.into(),
+        }
+    }
+}
+
+/// A struct representing ticker info for an asset pair including price
+/// information.
+///
+/// Note that this is only a small subset of the data provided by the Kraken API
+/// and only the parts required for retrieving price estimates for the solver
+/// are included.
+#[derive(Clone, Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct TickerInfo {
+    pub p: PricePair,
+}
+
+impl TickerInfo {
+    /// Create a new ticker info from its price pair.
+    #[cfg(test)]
+    pub fn new(today: f64, last_24h: f64) -> TickerInfo {
+        TickerInfo {
+            p: PricePair(today, last_24h),
+        }
+    }
+}
+
+/// A price pair used in the ticker info, where the first field is today's price
+/// and the second field is from the last 24 hours.
+#[derive(Copy, Clone, Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct PricePair(
+    #[serde(with = "display_fromstr")] pub f64,
+    #[serde(with = "display_fromstr")] pub f64,
+);
+
+impl PricePair {
+    /// Retrieves the price since midnight.
```

Comment:
Is this the price "at midnight" because prices fluctuate "since" a point in time and I would expect to see more than just a single value here... same question for last 24 hours.

Reply:
From what I understood from the documentation price pairs always contain two values, one which is "today", or since 00:00, and the other which is for the "last 24h". Depending on what the price pair is for, it can mean different things. So the ticker info JSON that is retrieved from Kraken, it contains things like the low, high and weighted average prices and for each of these, the price pair respectively represents the low price since 00:00 and the low price in the last 24h, the high price since 00:00 and the high price in the last 24h, and the weighted average price since 00:00 and in the last 24h.